// @tanstack/devtools-utils/dist/solid/esm/index.js
import { createComponent, template, insert, use } from 'solid-js/web';
import { createSignal, onMount, onCleanup } from 'solid-js';

// src/solid/class.tsx
var _tmpl$ = /* @__PURE__ */ template(`<div>`);
function constructCoreClass(Component) {
  class DevtoolsCore {
    #isMounted = false;
    #dispose;
    #Component;
    #ThemeProvider;
    constructor() {
    }
    async mount(el, theme) {
      const {
        lazy
      } = await import('solid-js');
      const {
        render,
        Portal
      } = await import('solid-js/web');
      if (this.#isMounted) {
        throw new Error("Devtools is already mounted");
      }
      const mountTo = el;
      const dispose = render(() => {
        this.#Component = Component;
        this.#ThemeProvider = lazy(() => import('@tanstack/devtools-ui').then((mod) => ({
          default: mod.ThemeContextProvider
        })));
        const Devtools = this.#Component;
        const ThemeProvider = this.#ThemeProvider;
        return createComponent(Portal, {
          mount: mountTo,
          get children() {
            var _el$ = _tmpl$();
            _el$.style.setProperty("height", "100%");
            insert(_el$, createComponent(ThemeProvider, {
              theme,
              get children() {
                return createComponent(Devtools, {});
              }
            }));
            return _el$;
          }
        });
      }, mountTo);
      this.#isMounted = true;
      this.#dispose = dispose;
    }
    unmount() {
      if (!this.#isMounted) {
        throw new Error("Devtools is not mounted");
      }
      this.#dispose?.();
      this.#isMounted = false;
    }
  }
  class NoOpDevtoolsCore extends DevtoolsCore {
    constructor() {
      super();
    }
    async mount(_el, _theme) {
    }
    unmount() {
    }
  }
  return [DevtoolsCore, NoOpDevtoolsCore];
}
var _tmpl$2 = /* @__PURE__ */ template(`<div>`);
function createSolidPanel(CoreClass) {
  function Panel(props) {
    let devToolRef;
    const [devtools] = createSignal(new CoreClass());
    onMount(() => {
      if (devToolRef) {
        devtools().mount(devToolRef, props?.theme ?? "dark");
      }
      onCleanup(() => {
        devtools().unmount();
      });
    });
    return (() => {
      var _el$ = _tmpl$2();
      var _ref$ = devToolRef;
      typeof _ref$ === "function" ? use(_ref$, _el$) : devToolRef = _el$;
      _el$.style.setProperty("height", "100%");
      return _el$;
    })();
  }
  function NoOpPanel(_props) {
    return [];
  }
  return [Panel, NoOpPanel];
}
function createSolidPlugin({
  Component,
  ...config
}) {
  function Plugin() {
    return {
      ...config,
      render: (_el, theme) => {
        return createComponent(Component, {
          theme
        });
      }
    };
  }
  function NoOpPlugin() {
    return {
      ...config,
      render: (_el, _theme) => []
    };
  }
  return [Plugin, NoOpPlugin];
}

export { constructCoreClass, createSolidPanel, createSolidPlugin };
