// @tanstack/devtools-utils/dist/solid/esm/server.js
import { createComponent, ssr, escape } from 'solid-js/web';
import { createSignal, onMount, onCleanup } from 'solid-js';

// src/solid/class.tsx
var _tmpl$ = ['<div style="', '">', "</div>"];
function constructCoreClass(Component) {
  class DevtoolsCore {
    #isMounted = false;
    #dispose;
    #Component;
    #ThemeProvider;
    constructor() {
    }
    async mount(el, theme) {
      const {
        lazy
      } = await import('solid-js');
      const {
        render,
        Portal
      } = await import('solid-js/web');
      if (this.#isMounted) {
        throw new Error("Devtools is already mounted");
      }
      const mountTo = el;
      const dispose = render(() => {
        this.#Component = Component;
        this.#ThemeProvider = lazy(() => import('@tanstack/devtools-ui').then((mod) => ({
          default: mod.ThemeContextProvider
        })));
        const Devtools = this.#Component;
        const ThemeProvider = this.#ThemeProvider;
        return createComponent(Portal, {
          mount: mountTo,
          get children() {
            return ssr(_tmpl$, "height:100%", escape(createComponent(ThemeProvider, {
              theme,
              get children() {
                return createComponent(Devtools, {});
              }
            })));
          }
        });
      }, mountTo);
      this.#isMounted = true;
      this.#dispose = dispose;
    }
    unmount() {
      if (!this.#isMounted) {
        throw new Error("Devtools is not mounted");
      }
      this.#dispose?.();
      this.#isMounted = false;
    }
  }
  class NoOpDevtoolsCore extends DevtoolsCore {
    constructor() {
      super();
    }
    async mount(_el, _theme) {
    }
    unmount() {
    }
  }
  return [DevtoolsCore, NoOpDevtoolsCore];
}
var _tmpl$2 = ['<div style="', '"></div>'];
function createSolidPanel(CoreClass) {
  function Panel(props) {
    const [devtools] = createSignal(new CoreClass());
    onMount(() => {
      onCleanup(() => {
        devtools().unmount();
      });
    });
    return ssr(_tmpl$2, "height:100%");
  }
  function NoOpPanel(_props) {
    return [];
  }
  return [Panel, NoOpPanel];
}
function createSolidPlugin({
  Component,
  ...config
}) {
  function Plugin() {
    return {
      ...config,
      render: (_el, theme) => {
        return createComponent(Component, {
          theme
        });
      }
    };
  }
  function NoOpPlugin() {
    return {
      ...config,
      render: (_el, _theme) => []
    };
  }
  return [Plugin, NoOpPlugin];
}

export { constructCoreClass, createSolidPanel, createSolidPlugin };
