// @tanstack/form-devtools/dist/esm/components/ActionButtons.js
import { template, effect, className, delegateEvents } from "solid-js/web";
import { formEventClient } from "@tanstack/form-core";
import { useStyles } from "../styles/use-styles.js";
var _tmpl$ = /* @__PURE__ */ template(`<div><button><span></span>Flush</button><button><span></span>Reset</button><button><span></span>Submit (-f)`);
function ActionButtons(props) {
  const styles = useStyles();
  return (() => {
    var _el$ = _tmpl$(), _el$2 = _el$.firstChild, _el$3 = _el$2.firstChild, _el$4 = _el$2.nextSibling, _el$5 = _el$4.firstChild, _el$6 = _el$4.nextSibling, _el$7 = _el$6.firstChild;
    _el$2.$$mousedown = () => {
      formEventClient.emit("request-form-state", {
        id: props.selectedInstance()?.id
      });
    };
    _el$4.$$mousedown = () => {
      formEventClient.emit("request-form-reset", {
        id: props.selectedInstance()?.id
      });
    };
    _el$6.$$mousedown = () => {
      formEventClient.emit("request-form-force-submit", {
        id: props.selectedInstance()?.id
      });
    };
    effect((_p$) => {
      var _v$ = styles().actionsRow, _v$2 = styles().actionButton, _v$3 = styles().actionDotGreen, _v$4 = styles().actionButton, _v$5 = styles().actionDotRed, _v$6 = styles().actionButton, _v$7 = styles().actionDotYellow;
      _v$ !== _p$.e && className(_el$, _p$.e = _v$);
      _v$2 !== _p$.t && className(_el$2, _p$.t = _v$2);
      _v$3 !== _p$.a && className(_el$3, _p$.a = _v$3);
      _v$4 !== _p$.o && className(_el$4, _p$.o = _v$4);
      _v$5 !== _p$.i && className(_el$5, _p$.i = _v$5);
      _v$6 !== _p$.n && className(_el$6, _p$.n = _v$6);
      _v$7 !== _p$.s && className(_el$7, _p$.s = _v$7);
      return _p$;
    }, {
      e: void 0,
      t: void 0,
      a: void 0,
      o: void 0,
      i: void 0,
      n: void 0,
      s: void 0
    });
    return _el$;
  })();
}
delegateEvents(["mousedown"]);
export {
  ActionButtons
};
