// @tanstack/form-devtools/dist/esm/components/DetailsPanel.js
import { template, insert, createComponent, effect, className } from "solid-js/web";
import { createMemo, Show, For } from "solid-js";
import { JsonTree } from "@tanstack/devtools-ui";
import { useStyles } from "../styles/use-styles.js";
import { useFormEventClient } from "../contexts/eventClientContext.js";
import { ActionButtons } from "./ActionButtons.js";
import { StateHeader } from "./StateHeader.js";
var _tmpl$ = /* @__PURE__ */ template(`<div><div><div>Actions</div></div><div><div>Individual Fields</div><div></div></div><div><div>Form values</div><div></div></div><div><div>Form status</div><div></div></div><div><div>Form options</div><div></div></div><div><div>Submission history</div><div>`), _tmpl$2 = /* @__PURE__ */ template(`<div>`), _tmpl$3 = /* @__PURE__ */ template(`<div><div>`);
function DetailsPanel({
  selectedKey
}) {
  const styles = useStyles();
  const {
    store
  } = useFormEventClient();
  const selectedIndex = createMemo(() => store().findIndex((f) => f.id === selectedKey()));
  const selectedInstance = createMemo(() => selectedIndex() > -1 ? store()[selectedIndex()] : null);
  const state = createMemo(() => selectedInstance()?.state);
  const formStatus = createMemo(() => ({
    canSubmit: state()?.canSubmit,
    isFormValid: state()?.isFormValid,
    isFormValidating: state()?.isFormValidating,
    isSubmitted: state()?.isSubmitted,
    isSubmitting: state()?.isSubmitting,
    isSubmitSuccessful: state()?.isSubmitSuccessful,
    submissionAttempts: state()?.submissionAttempts,
    errors: state()?.errors,
    errorMap: state()?.errorMap
  }));
  const individualFields = createMemo(() => {
    const fields = {};
    const values = state()?.values || {};
    const fieldMeta = state()?.fieldMeta || {};
    Object.keys(values).forEach((key) => {
      fields[key] = {
        value: values[key],
        meta: fieldMeta[key]
      };
    });
    return fields;
  });
  return (() => {
    var _el$ = _tmpl$2();
    insert(_el$, createComponent(Show, {
      get when() {
        return selectedInstance();
      },
      get children() {
        return [createComponent(StateHeader, {
          selectedInstance
        }), (() => {
          var _el$2 = _tmpl$(), _el$3 = _el$2.firstChild, _el$4 = _el$3.firstChild, _el$5 = _el$3.nextSibling, _el$6 = _el$5.firstChild, _el$7 = _el$6.nextSibling, _el$8 = _el$5.nextSibling, _el$9 = _el$8.firstChild, _el$0 = _el$9.nextSibling, _el$1 = _el$8.nextSibling, _el$10 = _el$1.firstChild, _el$11 = _el$10.nextSibling, _el$12 = _el$1.nextSibling, _el$13 = _el$12.firstChild, _el$14 = _el$13.nextSibling, _el$15 = _el$12.nextSibling, _el$16 = _el$15.firstChild, _el$17 = _el$16.nextSibling;
          insert(_el$3, createComponent(ActionButtons, {
            selectedInstance
          }), null);
          _el$7.style.setProperty("display", "flex");
          _el$7.style.setProperty("gap", "8px");
          insert(_el$7, createComponent(For, {
            get each() {
              return Object.entries(individualFields());
            },
            children: ([fieldName, fieldData]) => (() => {
              var _el$18 = _tmpl$3(), _el$19 = _el$18.firstChild;
              _el$18.style.setProperty("margin-bottom", "16px");
              _el$19.style.setProperty("font-weight", "bold");
              _el$19.style.setProperty("margin-bottom", "4px");
              insert(_el$19, fieldName);
              insert(_el$18, createComponent(JsonTree, {
                copyable: true,
                value: fieldData
              }), null);
              effect(() => className(_el$18, styles().stateContent));
              return _el$18;
            })()
          }));
          insert(_el$0, createComponent(JsonTree, {
            copyable: true,
            get value() {
              return store()[selectedIndex()].state.values;
            }
          }));
          insert(_el$11, createComponent(JsonTree, {
            copyable: true,
            get value() {
              return formStatus();
            }
          }));
          insert(_el$14, createComponent(JsonTree, {
            copyable: true,
            get value() {
              return store()[selectedIndex()]?.options;
            },
            collapsePaths: ["validators"]
          }));
          insert(_el$17, createComponent(JsonTree, {
            copyable: true,
            get value() {
              return store()[selectedIndex()]?.history;
            }
          }));
          effect((_p$) => {
            var _v$ = styles().detailsGrid, _v$2 = styles().detailSection, _v$3 = styles().detailSectionHeader, _v$4 = styles().detailSection, _v$5 = styles().detailSectionHeader, _v$6 = styles().detailSection, _v$7 = styles().detailSectionHeader, _v$8 = styles().stateContent, _v$9 = styles().detailSection, _v$0 = styles().detailSectionHeader, _v$1 = styles().stateContent, _v$10 = styles().detailSection, _v$11 = styles().detailSectionHeader, _v$12 = styles().stateContent, _v$13 = styles().detailSection, _v$14 = styles().detailSectionHeader, _v$15 = styles().stateContent;
            _v$ !== _p$.e && className(_el$2, _p$.e = _v$);
            _v$2 !== _p$.t && className(_el$3, _p$.t = _v$2);
            _v$3 !== _p$.a && className(_el$4, _p$.a = _v$3);
            _v$4 !== _p$.o && className(_el$5, _p$.o = _v$4);
            _v$5 !== _p$.i && className(_el$6, _p$.i = _v$5);
            _v$6 !== _p$.n && className(_el$8, _p$.n = _v$6);
            _v$7 !== _p$.s && className(_el$9, _p$.s = _v$7);
            _v$8 !== _p$.h && className(_el$0, _p$.h = _v$8);
            _v$9 !== _p$.r && className(_el$1, _p$.r = _v$9);
            _v$0 !== _p$.d && className(_el$10, _p$.d = _v$0);
            _v$1 !== _p$.l && className(_el$11, _p$.l = _v$1);
            _v$10 !== _p$.u && className(_el$12, _p$.u = _v$10);
            _v$11 !== _p$.c && className(_el$13, _p$.c = _v$11);
            _v$12 !== _p$.w && className(_el$14, _p$.w = _v$12);
            _v$13 !== _p$.m && className(_el$15, _p$.m = _v$13);
            _v$14 !== _p$.f && className(_el$16, _p$.f = _v$14);
            _v$15 !== _p$.y && className(_el$17, _p$.y = _v$15);
            return _p$;
          }, {
            e: void 0,
            t: void 0,
            a: void 0,
            o: void 0,
            i: void 0,
            n: void 0,
            s: void 0,
            h: void 0,
            r: void 0,
            d: void 0,
            l: void 0,
            u: void 0,
            c: void 0,
            w: void 0,
            m: void 0,
            f: void 0,
            y: void 0
          });
          return _el$2;
        })()];
      }
    }));
    effect(() => className(_el$, styles().stateDetails));
    return _el$;
  })();
}
export {
  DetailsPanel
};
