// @tanstack/form-devtools/dist/esm/components/Shell.js
import { createComponent, template, insert, effect, className, delegateEvents } from "solid-js/web";
import { createSignal, onMount, onCleanup, Show } from "solid-js";
import { MainPanel, Header, HeaderLogo } from "@tanstack/devtools-ui";
import { useStyles } from "../styles/use-styles.js";
import { UtilList } from "./UtilList.js";
import { DetailsPanel } from "./DetailsPanel.js";
var _tmpl$ = /* @__PURE__ */ template(`<div>Details`), _tmpl$2 = /* @__PURE__ */ template(`<div><div></div><div></div><div>`);
function Shell() {
  const styles = useStyles();
  const [leftPanelWidth, setLeftPanelWidth] = createSignal(300);
  const [isDragging, setIsDragging] = createSignal(false);
  const [selectedKey, setSelectedKey] = createSignal(null);
  let dragStartX = 0;
  let dragStartWidth = 0;
  const handleMouseDown = (e) => {
    e.preventDefault();
    e.stopPropagation();
    setIsDragging(true);
    document.body.style.cursor = "col-resize";
    document.body.style.userSelect = "none";
    dragStartX = e.clientX;
    dragStartWidth = leftPanelWidth();
  };
  const handleMouseMove = (e) => {
    if (!isDragging()) return;
    e.preventDefault();
    const deltaX = e.clientX - dragStartX;
    const newWidth = Math.max(150, Math.min(800, dragStartWidth + deltaX));
    setLeftPanelWidth(newWidth);
  };
  const handleMouseUp = () => {
    setIsDragging(false);
    document.body.style.cursor = "";
    document.body.style.userSelect = "";
  };
  onMount(() => {
    document.addEventListener("mousemove", handleMouseMove);
    document.addEventListener("mouseup", handleMouseUp);
  });
  onCleanup(() => {
    document.removeEventListener("mousemove", handleMouseMove);
    document.removeEventListener("mouseup", handleMouseUp);
  });
  return createComponent(MainPanel, {
    get children() {
      return [createComponent(Header, {
        get children() {
          return createComponent(HeaderLogo, {
            flavor: {
              light: "#eeaf00",
              dark: "#eeaf00"
            },
            onClick: () => {
              window.open("https://tanstack.com/form/latest/docs/overview", "_blank");
            },
            children: "TanStack Form"
          });
        }
      }), (() => {
        var _el$ = _tmpl$2(), _el$2 = _el$.firstChild, _el$3 = _el$2.nextSibling, _el$4 = _el$3.nextSibling;
        _el$2.style.setProperty("min-width", "150px");
        _el$2.style.setProperty("max-width", "800px");
        insert(_el$2, createComponent(UtilList, {
          selectedKey,
          setSelectedKey
        }));
        _el$3.$$mousedown = handleMouseDown;
        _el$4.style.setProperty("flex", "1");
        insert(_el$4, createComponent(Show, {
          get when() {
            return selectedKey() != null;
          },
          get children() {
            return [(() => {
              var _el$5 = _tmpl$();
              effect(() => className(_el$5, styles().panelHeader));
              return _el$5;
            })(), createComponent(DetailsPanel, {
              selectedKey
            })];
          }
        }));
        effect((_p$) => {
          var _v$ = styles().mainContainer, _v$2 = styles().leftPanel, _v$3 = `${leftPanelWidth()}px`, _v$4 = `${styles().dragHandle} ${isDragging() ? "dragging" : ""}`, _v$5 = styles().rightPanel;
          _v$ !== _p$.e && className(_el$, _p$.e = _v$);
          _v$2 !== _p$.t && className(_el$2, _p$.t = _v$2);
          _v$3 !== _p$.a && ((_p$.a = _v$3) != null ? _el$2.style.setProperty("width", _v$3) : _el$2.style.removeProperty("width"));
          _v$4 !== _p$.o && className(_el$3, _p$.o = _v$4);
          _v$5 !== _p$.i && className(_el$4, _p$.i = _v$5);
          return _p$;
        }, {
          e: void 0,
          t: void 0,
          a: void 0,
          o: void 0,
          i: void 0
        });
        return _el$;
      })()];
    }
  });
}
delegateEvents(["mousedown"]);
export {
  Shell
};
