// @tanstack/form-devtools/dist/esm/components/StateHeader.js
import { template, insert, effect, className } from "solid-js/web";
import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime.js";
import { createSignal, onMount, onCleanup, createMemo } from "solid-js";
import { useStyles } from "../styles/use-styles.js";
var _tmpl$ = /* @__PURE__ */ template(`<div><div>Form state</div><div><div><div>Key</div><div></div><div>Last Updated</div><div> (<!>)`);
dayjs.extend(relativeTime);
function StateHeader(props) {
  const styles = useStyles();
  const [now, setNow] = createSignal(dayjs().unix());
  onMount(() => {
    const interval = setInterval(() => setNow(dayjs().unix()), 1e3);
    onCleanup(() => clearInterval(interval));
  });
  const state = props.selectedInstance;
  const updatedAt = createMemo(() => state()?.date.unix() ?? dayjs().unix());
  const relativeTimeText = createMemo(() => {
    const diffSeconds = Math.max(now() - updatedAt(), 0);
    if (diffSeconds < 60) {
      return `${diffSeconds} second${diffSeconds !== 1 ? "s" : ""} ago`;
    }
    return dayjs.unix(updatedAt()).fromNow();
  });
  if (!state()) return null;
  return (() => {
    var _el$ = _tmpl$(), _el$2 = _el$.firstChild, _el$3 = _el$2.nextSibling, _el$4 = _el$3.firstChild, _el$5 = _el$4.firstChild, _el$6 = _el$5.nextSibling, _el$7 = _el$6.nextSibling, _el$8 = _el$7.nextSibling, _el$9 = _el$8.firstChild, _el$1 = _el$9.nextSibling;
    _el$1.nextSibling;
    _el$3.style.setProperty("display", "flex");
    _el$3.style.setProperty("align-items", "center");
    _el$3.style.setProperty("gap", "16px");
    insert(_el$6, () => state().id);
    insert(_el$8, () => new Date(updatedAt() * 1e3).toLocaleTimeString(), _el$9);
    insert(_el$8, relativeTimeText, _el$1);
    effect((_p$) => {
      var _v$ = styles().stateHeader, _v$2 = styles().stateTitle, _v$3 = styles().infoGrid, _v$4 = styles().infoLabel, _v$5 = styles().infoValueMono, _v$6 = styles().infoLabel, _v$7 = styles().infoValueMono;
      _v$ !== _p$.e && className(_el$, _p$.e = _v$);
      _v$2 !== _p$.t && className(_el$2, _p$.t = _v$2);
      _v$3 !== _p$.a && className(_el$4, _p$.a = _v$3);
      _v$4 !== _p$.o && className(_el$5, _p$.o = _v$4);
      _v$5 !== _p$.i && className(_el$6, _p$.i = _v$5);
      _v$6 !== _p$.n && className(_el$7, _p$.n = _v$6);
      _v$7 !== _p$.s && className(_el$8, _p$.s = _v$7);
      return _p$;
    }, {
      e: void 0,
      t: void 0,
      a: void 0,
      o: void 0,
      i: void 0,
      n: void 0,
      s: void 0
    });
    return _el$;
  })();
}
export {
  StateHeader
};
