// @tanstack/form-devtools/dist/esm/styles/tokens.js
const tokens = {
  colors: {
    black: "#000000",
    white: "#ffffff",
    darkGray: {
      400: "#313749",
      500: "#292e3d",
      600: "#212530",
      700: "#191c24",
      800: "#111318"
    },
    gray: {
      100: "#f2f4f7",
      200: "#eaecf0",
      300: "#d0d5dd",
      400: "#98a2b3",
      500: "#667085",
      600: "#475467",
      800: "#1d2939",
      900: "#101828"
    },
    blue: {
      100: "#D1E9FF",
      300: "#84CAFF",
      400: "#53B1FD",
      500: "#2E90FA",
      600: "#1570EF",
      700: "#175CD3",
      900: "#194185"
    },
    green: {
      400: "#32D583"
    },
    red: {
      400: "#f87171"
    },
    yellow: {
      400: "#FDB022"
    },
    purple: {
      400: "#9B8AFB"
    },
    pink: {
      400: "#ec4899"
    }
  },
  alpha: {
    80: "cc",
    30: "4d",
    20: "33"
  },
  font: {
    size: {
      xs: "calc(var(--tsrd-font-size) * 0.75)",
      sm: "calc(var(--tsrd-font-size) * 0.875)",
      md: "var(--tsrd-font-size)",
      lg: "calc(var(--tsrd-font-size) * 1.125)"
    },
    weight: {
      semibold: "600",
      bold: "700"
    },
    fontFamily: {
      mono: `ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, 'Liberation Mono', 'Courier New', monospace`
    }
  },
  border: {
    radius: {
      sm: "calc(var(--tsrd-font-size) * 0.25)",
      md: "calc(var(--tsrd-font-size) * 0.375)",
      lg: "calc(var(--tsrd-font-size) * 0.5)"
    }
  },
  size: {
    1: "calc(var(--tsrd-font-size) * 0.25)",
    2: "calc(var(--tsrd-font-size) * 0.5)",
    3: "calc(var(--tsrd-font-size) * 0.75)",
    4: "calc(var(--tsrd-font-size) * 1)"
  },
  shadow: {
    xs: (_ = "rgb(0 0 0 / 0.1)") => `0 1px 2px 0 rgb(0 0 0 / 0.05)`,
    sm: (color = "rgb(0 0 0 / 0.1)") => `0 1px 3px 0 ${color}, 0 1px 2px -1px ${color}`,
    md: (color = "rgb(0 0 0 / 0.1)") => `0 4px 6px -1px ${color}, 0 2px 4px -2px ${color}`,
    lg: (color = "rgb(0 0 0 / 0.1)") => `0 10px 15px -3px ${color}, 0 4px 6px -4px ${color}`,
    xl: (color = "rgb(0 0 0 / 0.1)") => `0 20px 25px -5px ${color}, 0 8px 10px -6px ${color}`,
    "2xl": (color = "rgb(0 0 0 / 0.25)") => `0 25px 50px -12px ${color}`,
    inner: (color = "rgb(0 0 0 / 0.05)") => `inset 0 2px 4px 0 ${color}`,
    none: () => `none`
  }
};
export {
  tokens
};
