// @tanstack/form-devtools/dist/esm/contexts/eventClientContext.js
import { createComponent } from "solid-js/web";
import dayjs from "dayjs";
import { createContext, createEffect, onCleanup, useContext, createMemo } from "solid-js";
import { createStore } from "solid-js/store";
import { formEventClient } from "@tanstack/form-core";
function useProviderValue() {
  const [store, setStore] = createStore([]);
  createEffect(() => {
    const cleanup = formEventClient.on("form-api", (e) => {
      const id = e.payload.id;
      const existingIndex = store.findIndex((item) => item.id === id);
      if (existingIndex > -1) {
        setStore(existingIndex, {
          state: e.payload.state,
          options: e.payload.options,
          date: dayjs()
        });
      } else {
        setStore((prev) => [...prev, {
          id,
          state: e.payload.state,
          options: e.payload.options,
          date: dayjs(),
          history: []
        }]);
      }
    });
    onCleanup(cleanup);
  });
  createEffect(() => {
    const cleanup = formEventClient.on("form-state", (e) => {
      const id = e.payload.id;
      const existingIndex = store.findIndex((item) => item.id === id);
      if (existingIndex > -1) {
        setStore(existingIndex, {
          state: e.payload.state,
          date: dayjs()
        });
      } else {
        setStore((prev) => [...prev, {
          id,
          state: e.payload.state,
          options: {},
          date: dayjs(),
          history: []
        }]);
      }
    });
    onCleanup(cleanup);
  });
  createEffect(() => {
    const cleanup = formEventClient.on("form-submission", (e) => {
      const id = e.payload.id;
      const existingIndex = store.findIndex((item) => item.id === id);
      if (existingIndex > -1 && store[existingIndex]) {
        const {
          id: _,
          ...rest
        } = e.payload;
        const newHistory = [rest, ...store[existingIndex].history].slice(0, 5);
        setStore(existingIndex, "history", newHistory);
      }
    });
    onCleanup(cleanup);
  });
  createEffect(() => {
    const cleanup = formEventClient.on("form-unmounted", (e) => {
      setStore((prev) => prev.filter((item) => item.id !== e.payload.id));
    });
    onCleanup(cleanup);
  });
  return {
    store
  };
}
const FormEventClientContext = createContext(void 0);
const FormEventClientProvider = (props) => {
  const value = useProviderValue();
  return createComponent(FormEventClientContext.Provider, {
    value,
    get children() {
      return props.children;
    }
  });
};
function useFormEventClient() {
  const context = useContext(FormEventClientContext);
  if (context === void 0) {
    throw new Error(`useFormEventClient must be used within a FormEventClientContext`);
  }
  const memoContext = createMemo(() => context.store);
  return {
    store: memoContext
  };
}
export {
  FormEventClientProvider,
  useFormEventClient
};
